pyro/poutine/enumerate_poutine.py

```diff
@@ -0,0 +1,68 @@
+from __future__ import absolute_import, division, print_function
+
+from .poutine import Messenger, Poutine
+
+
+def _iter_discrete_filter(msg):
+    return ((msg["type"] == "sample") and
+            (not msg["is_observed"]) and
+            (msg["infer"].get("enumerate") == "parallel"))
+
+
+class EnumerateMessenger(Messenger):
+    """
+    Enumerates in parallel over discrete sample sites marked
+    ``infer={"enumerate": "parallel"}``.
+
+    :param int first_available_dim: The first tensor dimension (counting
+        from the right) that is available for parallel enumeration. This
+        dimension and all dimensions left may be used internally by Pyro.
+    """
+    def __init__(self, first_available_dim):
+        super(EnumerateMessenger, self).__init__()
+        self.next_available_dim = first_available_dim
+
+    def _pyro_sample(self, msg):
+        """
+        :param msg: current message at a trace site.
+        :returns: a sample from the stochastic function at the site.
+        """
+        if not msg["done"] and _iter_discrete_filter(msg):
+            value = msg["fn"].enumerate_support()
+
+            # Ensure enumeration happens at an available tensor dimension.
+            event_dim = len(msg["fn"].event_shape)
+            actual_dim = value.dim() - event_dim - 1
```

Comment:
can we have one or two comments here i find this confusing
```
actual_dim = value.dim() - event_dim - 1
```

Reply:
Ok, I've simplified and added some comments.